### ISSUE TITLE
Move pre-calculated DSP data from SPARQL to PostgreSQL

### ISSUE THREAD

**Issue by siuc-nate** [CredentialEngine/CredentialRegistry#342]
Due to the poor performance of SPARQL for selecting this data, and the language limitation that prevents selecting only the first x items from each related item set, we should move the pre-calculated description set data from SPARQL to the PostgreSQL data and enable accessing it via API.

As a refresher, the pre-calculated DSP data looks like this (with the addition of a count of total URIs):
https://github.com/CredentialEngine/CredentialRegistry/issues/322
```
[
  {
    "@id":"_:abcdefg...",
    "credreg:__dspPath": "> ceterms:requires > ceterms:ConditionProfile > ceterms:targetLearningOpportunity > ceterms:LearningOpportunityProfile",
    "credreg:__dspQuery": "?subject ( ceterms:requires/ceterms:targetLearningOpportunity/^credreg:__graph? ) ?target .",
    "credreg:__dspURI": [
      "https://.../resources/ce-abc...",
      "https://.../resources/ce-def...",
      "https://.../resources/ce-ghi..."      
    ],
    "credreg:__dspURITotal": 3
  },
  {
    "@id":"_:hijklm...",
    "credreg:__dspPath": "< ceterms:targetCredential < ceterms:ConditionProfile < ceterms:isRequiredFor < ceterms:LearningOpportunityProfile",
    "credreg:__dspQuery": "?subject ( credreg:__graph?/^ceterms:targetCredential/^ceterms:isRequiredFor ) ?target . ?target a ceterms:LearningOpportunityProfile .",
    "credreg:__dspURI": [
      "https://.../resources/ce-abc...",
      "https://.../resources/ce-ghi..."
    ],
    "credreg:__dspURITotal": 2
  },
  {
    "@id":"_:nopqrst...",
    "credreg:__dspPath": "> ceterms:requires > ceterms:ConditionProfile > ceterms:targetLearningOpportunity > ceterms:LearningOpportunityProfile > ceterms:teaches > ceterms:CredentialAlignmentObject > ceterms:targetNode > ceasn:Competency",
    "credreg:__dspQuery": "?subject ( ceterms:requires/ceterms:targetLearningOpportunity/^credreg:__graph?/ceterms:teaches/ceterms:targetNode ) ?target .",
    "credreg:__dspURI": [
      "https://.../resources/ce-jkl...",
      "https://.../resources/ce-mno...",
      "https://.../resources/ce-pqr...",
      "https://.../resources/ce-stu...",
      "https://.../resources/ce-vwx..."    
    ],
    "credreg:__dspURITotal": 5
  }
]
```

Only paths with at least one `credreg:__dspURI` need to be included/stored.  It's also not necessary to return (and probably not to store, unless it's useful internally) the value for `credreg:__dspQuery`.  It may not be necessary to return the value of `@id` either, since this would be implicit from the requested CTID, but we can discuss this.

A sensible API endpoint would probably be:
`https://credentialengineregistry.org/descriptionset/ce-b1fda25e-2e21-4a71-8cdd-d608892bcfe4`
With parameters like:
- `perPathLimit`, which would take an integer and control how many values for each `credreg:__dspURI` are returned for each path. By default, no limit would be applied.
- `pathContains`, which would take a string and return only those items with a `credreg:__dspPath` that contains the indicated path (to allow for partial matches). By default, no constraint would be applied.
- `pathExact`, which would do the same thing, but only return data where the `credreg:__dspPath` is an exact match for the provided string. By default, no constraint would be applied.

**Example request/response:**
Request:
`https://credentialengineregistry.org/descriptionset/ce-b1fda25e-2e21-4a71-8cdd-d608892bcfe4`
(probably with an API key in the headers somewhere)
Response:
```
{
  "credreg:__relatedItemsMap": [
    {
      "credreg:__dspPath": "> ceterms:requires > ceterms:ConditionProfile > ceterms:targetLearningOpportunity > ceterms:LearningOpportunityProfile",
      "credreg:__dspURI": [
        "https://.../resources/ce-abc...",
        "https://.../resources/ce-def...",
        "https://.../resources/ce-ghi..."      
      ],
      "credreg:__dspURITotal": 3
    },
    {
      "credreg:__dspPath": "< ceterms:targetCredential < ceterms:ConditionProfile < ceterms:isRequiredFor < ceterms:LearningOpportunityProfile",
      "credreg:__dspURI": [
        "https://.../resources/ce-abc...",
        "https://.../resources/ce-ghi..."
      ],
      "credreg:__dspURITotal": 2
    },
    {
      "credreg:__dspPath": "> ceterms:requires > ceterms:ConditionProfile > ceterms:targetLearningOpportunity > ceterms:LearningOpportunityProfile > ceterms:teaches > ceterms:CredentialAlignmentObject > ceterms:targetNode > ceasn:Competency",
      "credreg:__dspURI": [
        "https://.../resources/ce-jkl...",
        "https://.../resources/ce-mno...",
        "https://.../resources/ce-pqr...",
        "https://.../resources/ce-stu...",
        "https://.../resources/ce-vwx...",
        "https://.../resources/ce-vwx...",
        "https://.../resources/ce-yz1...",
        "https://.../resources/ce-234...",
        "https://.../resources/ce-567...",
        "https://.../resources/ce-890...",
        "https://.../resources/ce-101..."    
      ],
      "credreg:__dspURITotal": 11
    },
    ... //Other paths
  ]
}
```
This is the basic "get everything" use case, and is more or less what we're currently doing through SPARQL.

Request:
`https://credentialengineregistry.org/descriptionset/ce-b1fda25e-2e21-4a71-8cdd-d608892bcfe4?perPathlimit=5&pathContains=%3E%20ceterms:requires`
(probably with an API key in the headers somewhere)
Response:
```
{
  "credreg:__relatedItemsMap": [
    {
      "credreg:__dspPath": "> ceterms:requires > ceterms:ConditionProfile > ceterms:targetLearningOpportunity > ceterms:LearningOpportunityProfile",
      "credreg:__dspURI": [
        "https://.../resources/ce-abc...",
        "https://.../resources/ce-def...",
        "https://.../resources/ce-ghi..."      
      ],
      "credreg:__dspURITotal": 3
    },
    {
      "credreg:__dspPath": "> ceterms:requires > ceterms:ConditionProfile > ceterms:targetLearningOpportunity > ceterms:LearningOpportunityProfile > ceterms:teaches > ceterms:CredentialAlignmentObject > ceterms:targetNode > ceasn:Competency",
      "credreg:__dspURI": [
        "https://.../resources/ce-jkl...",
        "https://.../resources/ce-mno...",
        "https://.../resources/ce-pqr...",
        "https://.../resources/ce-stu...",
        "https://.../resources/ce-vwx..."
      ],
      "credreg:__dspURITotal": 11
    },
    ... //Other matching paths
  ]
}
```
This allows me to get the first 5 URIs for every path that contains `> ceterms:requires`, while also being able to know that there are exactly 3 URIs and 11 URIs for those respective paths. A more likely version of this that would apply to the Finder would be to get the first 10 URIs for each path.

I'm open to pushback/feedback on this.  We probably don't need to use property names prefixed with `credreg:__`, for example.  

I also went back and forth a few times on whether or not it makes sense to enable including the actual related documents as part of this request, but since we have the multi-get by CTID now, I don't think we need to complicate the implementation of this with that.

**Comment by siuc-nate** [comment]
This is being worked on according to @edgarf in our meeting today.

**Comment by excelsior** [comment]
This feature is available on production.

The API looks like this:

```
https://credentialengineregistry.org/description_sets/ce-7fceca4b-722e-4885-b3ae-a9f40504815c?limit=10&path_contains=own
```

It simply returns an array of objects with the `path` and `uris` properties. The `limit` and `path_contains` are optional.

Path matching is case-insensitive, but the values are not normalized in any way (i.e. `foo>bar` won't match `foo > bar`). Also, things like colons, spaces, etc. need to be properly encoded.

I haven't fully integrated this code into the indexing mechanism, because the current implementation of description set pre-calculating in Postgres is terribly slow (takes ~8h for the complete dataset). I'll proceed with optimization once the requirements are finalized.

**Comment by siuc-nate** [comment]
Awesome, thanks.

Would it be faster to build the DSPs in SPARQL and store them in Postgres?

**Comment by siuc-nate** [comment]
This appears to be working now. A pull request should(?) close this issue automatically later.

**Comment by excelsior** [comment]
@siuc-nate I found a few invalid subqueries among the [DS item maps](https://docs.google.com/spreadsheets/d/1VNgJ7qu_zydCB3UQMzEuHbYKczr_Q9nkRRxzoUWYLuk/edit#gid=0): rows 38—45 contain an invalid `?credreg:__graph` path, should probably be `credreg:__graph?` or `^credreg:__graph?`. Could you please fix those, so I can pre-calculate the missing sets before we close this one.

**Comment by siuc-nate** [comment]
Good catch - I have updated the spreadsheet to change any rows with that error.  Let me know if you spot anything else.

**Comment by siuc-nate** [comment]
I have spotted a problem:

The DSP data does not appear to retrieve any bnode URIs, only full registry URIs.  This means, for instance, that the "accreditedBy" value for an organization like [NOCTI](https://credentialfinder.org/organization/1) comes back empty, even though there is a bnode that should be part of that description set.

For convenience:
`https://credentialengineregistry.org/description_sets/ce-6a62b250-a1a2-4d31-a702-cdc2437efd31?limit=10`

Related to this, it would be very useful if the existing search feature that allows getting multiple CTIDs at once was expanded to handle bnode URIs as well.  This would save multiple round-trips to the Registry downloading `@graph` data that is mostly not needed.

Either of the following would be useful POST bodies for `https://staging.credentialengineregistry.org/resources/search` (the second is probably cleaner):
```
{
  "ctids": [ "ce-00d41001-76b3-4fae-9508-0212b0cad347", "ce-03dc1c17-0a0e-47e7-b3d9-878f7ec03dfe", "ce-16ade8f0-16ee-4525-8964-1a87366c035c", "_:d4ff14cf-4a76-48bf-9250-f0df3fe3d278" ]
}
```
```
{
  "ctids": [ "ce-00d41001-76b3-4fae-9508-0212b0cad347", "ce-03dc1c17-0a0e-47e7-b3d9-878f7ec03dfe", "ce-16ade8f0-16ee-4525-8964-1a87366c035c" ],
  "bnodes": [ "_:d4ff14cf-4a76-48bf-9250-f0df3fe3d278" ]
}
```

**Comment by excelsior** [comment]
@siuc-nate I changed the pre-calculation algo to include bnodes (both as subjects and targets). I'll still need update single envelope processing, because it's based on extracted resources, but we don't extract blank node resources. Could you please confirm that the data looks good first?

**Comment by siuc-nate** [comment]
Thanks, but it doesn't appear to be working, at least not for the example in my previous post (unless I'm looking in the wrong place).

The graph for that has a bnode that isn't showing up.

**Comment by excelsior** [comment]
There are two problems here.

1. NOCTI is a Credential Organization, but there are no paths containing the `accreditedBy` property with `Credential Organization` as the starting point.
2. NOCTI's accreditor—ICAC—is a QA Credential Organization, and there are paths with `accreditedBy` for that. However, they a) are reverse and b) target a Credential, Learning Opportunity, or Assessment (rows 220, 224, and 228).

In other words, if there were a subquery for this relation, the DS data would've been available under

`https://credentialengineregistry.org/description_sets/_:d4ff14cf-4a76-48bf-9250-f0df3fe3d278`

**Comment by siuc-nate** [comment]
Good catch on the paths - it looks like a lot of the "accredits" relationships got missed.  I will make corrections.

**Comment by siuc-nate** [comment]
I've updated the spreadsheet:
- Used both organization classes (ceterms:CredentialOrganization and ceterms:QACredential Organization) for all organization starting points (column A). Previously some were just QA orgs, but we've encountered enough cases of organizations picking one but needing to do things from the other that we should just check both.
- Added rows for accredits, approves, recognizes, and regulates (and their reverse "___By" forms) as well as the _reverse_ forms of each
- Added [ceterms:CertificateOfCompletion](https://credreg.net/ctdl/terms/CertificateOfCompletion) to all of the `VALUES ?credentialType ...` parts of queries across the sheet
- Note: You will also need to add ceterms:CertificateOfCompletion to all of the "starting point" (column A) items that include Credential

**Comment by excelsior** [comment]
I've recalculated the description sets with the latest config. Now NOCTI's description sets contain bnode resources:

`https://credentialengineregistry.org/description_sets/ce-6a62b250-a1a2-4d31-a702-cdc2437efd31?limit=10&path_contains=accreditedBy`

**Comment by siuc-nate** [comment]
This is working now, and can be closed.
